### ISSUE TITLE
Bin packing algorithm causing high rate of false positives

### ISSUE THREAD

**Issue by danluu** [BitFunnel/BitFunnel#360]
In #358, we use the "standard" setup with adHocFrequency of 0.001.

For a density of 0.05 and the ClassicBitsliced treatment we get an SNR of 2.84. If we set adHocFrequency to 1.0, we get an SNR of 10.23. 

In the first setup, we had:

~~~
RowAssigner for rank 0
  Terms
    Total: 2105385
    Adhoc: 2093390
    Explicit: 11655
    Private: 340

  Rows
    Total: 7600
    Adhoc: 1842
    Shared Explicit: 5415
    Private Explicit: 340
    Explicit: 5758

  Bytes per document: 950.375


  Densities in explicit shared rows
    Mean: 0.0481016
    Min: 0.00100063
    Max: 0.0499995
    Variance: 5.26673e-05
~~~

In the second setup, we have:
~~~
RowAssigner for rank 0
  Terms
    Total: 2105385
    Adhoc: 2105385
    Explicit: 0
    Private: 0

  Rows
    Total: 5766
    Adhoc: 5763
    Shared Explicit: 0
    Private Explicit: 0
    Explicit: 3

  Bytes per document: 721.125


  Densities in explicit shared rows
    Mean: nan
    Min: nan
    Max: nan
    Variance: 1
~~~

The second, adHoc only treatment has better SNR *and* uses less space. It's entirely possible that we're better off with the adHocThreshold set to match the density, or something very similar. That would result in overdensity adHoc rows, but it would also cause rows that were previously packed into a few bins to get spread out over many more bins.

**Comment by danluu** [comment]
For reasons that will need some investigation, removing non-private explicitly packed rows increases the SNR for ClassicBitsliced and PrivatedSharedRank0 but makes the SNR for the Optimal treatment worse. The short-term fix for this is to only remove non-private explicit rows for the former treatment and not the latter.

The longer term fix is to understand what's going on here and then do the right thing.

**Comment by danluu** [comment]
The short-term fix for this got bundled into 02aeea63ac38a7280cf45c6e7178ccf3881b969d.
